Jetty 9 - Client-side problems with digest authentication

When a Jetty `HttpClient` sends a POST to a URI protected by a `DigestAuthenticator` on a Jetty server, and the server responds with an authentication challenge, the requests fails intermittently on the client side. The larger the content and the slower the internet connection, the more likely the failure occurs. The requests fails with randomly either an `org.eclipse.jetty.io.EofException` thrown in `ChannelEndPoint.flush` or a `java.io.EOFException` thrown in `HttpReceiverOverHTTP.earlyEOF`. Apparently, the client complains about the connection being closed by the server before the request content is sent completely.

The reasons why the server closes the connection without reading all the content are already discussed in issue #651. I also agree to the points raised by @gregw, but I think that the client should be ready for a server behaving in this way and attempt to recover from such a situation. This also seems to be in line with [RFC 7230, section 6.3.1](https://tools.ietf.org/html/rfc7230#section-6.3.1). In particular, since an authentication challenge response indicates that the request must be authenticated before the server can process the request and its content, the client could safely ignore the closed connection, follow the authentication protocol and resend the request with the established credentials using a new connection, even if the request uses a non-idempotent method such as POST.

I also agree with @sbordet that in the case of digest authentication, the credentials should be established with an initial request without content, but this does not solve the problem in the case of stale credentials of a later POST.

On the server side, I implemented @Adebski’s solution, which works fine apart from being a potential DOS risk as noted by @gregw. Therefore, I hope it is possible to change the `HttpClient` such that it also works with servers that close the connection without reading the whole content, at least in the case of digest authentication challenges.

By the way, I use the TCP/IP Monitor integrated in Eclipse IDE for network debugging. When I activate this monitor between a Jetty server and a Jetty `HttpClient`, requests that require digest authentication always fail, regardless of payload size or connection speed (even with client and server running on the same machine). Unfortunately, I cannot tell whether the problem is created by the client, the server, or the monitor, so I do not know where to raise an issue regarding this problem.

Our use case is the distribution and synchronization of data between multiple systems. So, our client executes POSTs, PUTs and DELETEs all day and only rarely a GET. Our exact test environment is Jetty 9.4.7.v20170914, Java 1.8.0_152 and Http1.1.

I think retrying the POST is only a problem if you use the `InputStreamContentProvider`
or an `AsyncContentProvider`. But for the others there is no problem because the  `iterator()` 
method always creates a new `Iterator`.
The `AuthenticationProtocolHandler.AuthenticationListener.onComplete()` already makes exactly this retry and copies the `ContentProvider` from the old `Request` into the new `Request`.
This is the behaviour I expect from the  `HttpClient`, and works fine on a lan, but failes if the connection is not fast enough.

I can show you my test scenario

This is my server
```java
import java.io.BufferedReader;
import java.io.IOException;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import org.eclipse.jetty.security.ConstraintMapping;
import org.eclipse.jetty.security.ConstraintSecurityHandler;
import org.eclipse.jetty.security.HashLoginService;
import org.eclipse.jetty.security.UserStore;
import org.eclipse.jetty.security.authentication.DigestAuthenticator;
import org.eclipse.jetty.util.security.Constraint;
import org.eclipse.jetty.util.security.Credential;

import org.eclipse.jetty.server.Request;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.handler.AbstractHandler;

public class HelloJetty extends AbstractHandler
{
    @Override
    public void handle( String target,
                        Request baseRequest,
                        HttpServletRequest request,
                        HttpServletResponse response ) throws IOException,
                                                      ServletException
    {
		String method = request.getMethod();
		if ("GET".equals(method)) {
			response.setContentType("text/html; charset=utf-8");
			response.setStatus(HttpServletResponse.SC_OK);
			response.getWriter().println("<h1>Hello Jetty</h1>");
		}
		else if ("POST".equals(method)) {
			BufferedReader reader = request.getReader();
			for (;;) {
				String line = reader.readLine();
				if (line == null)
					break;
				System.out.println(line);
			}
		}
        baseRequest.setHandled(true);
    }

    public static void main(String[] args) throws Exception
    {
        Server server = new Server(8080);

		HelloJetty helloJettyHandler = new HelloJetty();
		
		String realm = "HelloJettyRealm";
		String[] roles = new String[] { "user" };

		UserStore userStore = new UserStore();
		userStore.addUser("USERNAME", Credential.getCredential("PASSWORD"), roles);

		HashLoginService loginService = new HashLoginService();
		loginService.setUserStore(userStore);
		loginService.setName(realm);

		Constraint constraint = new Constraint();
		constraint.setName(Constraint.__DIGEST_AUTH);
		constraint.setRoles(roles);
		constraint.setAuthenticate(true);

		ConstraintMapping cm = new ConstraintMapping();
		cm.setConstraint(constraint);
		cm.setPathSpec("/");

		ConstraintSecurityHandler csHandler = new ConstraintSecurityHandler();
		csHandler.setAuthenticator(new DigestAuthenticator());
		csHandler.setRealmName(realm);
		csHandler.addConstraintMapping(cm);
		csHandler.setLoginService(loginService);
		csHandler.setHandler(helloJettyHandler);
		
        server.setHandler(csHandler);

        server.start();
        server.join();
    }
}
```

This is my client
```java
import java.net.URI;
import java.util.concurrent.TimeUnit;
import org.eclipse.jetty.client.HttpClient;
import org.eclipse.jetty.client.api.Authentication;
import org.eclipse.jetty.client.api.ContentResponse;
import org.eclipse.jetty.client.util.DigestAuthentication;
import org.eclipse.jetty.client.util.StringContentProvider;

public class HelloClient
{
    public static void main(String[] args) throws Exception
    {
		HttpClient httpClient = new HttpClient();
		try {
			String outStr =
"firstline\n" +
"line  2 ------------------------------------------------------------------ \n" +
"line  3 ------------------------------------------------------------------ \n" +
"line  4 ------------------------------------------------------------------ \n" +
"line  5 ------------------------------------------------------------------ \n" +
"line  6 ------------------------------------------------------------------ \n" +
"line  7 ------------------------------------------------------------------ \n" +
"line  8 ------------------------------------------------------------------ \n" +
"line  9 ------------------------------------------------------------------ \n" +
"line 10 ------------------------------------------------------------------ \n" +
"line 11 ------------------------------------------------------------------ \n" +
"line 12 ------------------------------------------------------------------ \n" +
"line 13 ------------------------------------------------------------------ \n" +
"line 14 ------------------------------------------------------------------ \n" +
"line 15 ------------------------------------------------------------------ \n" +
"line 16 ------------------------------------------------------------------ \n" +
"line 17 ------------------------------------------------------------------ \n" +
"line 18 ------------------------------------------------------------------ \n" +
"line 19 ------------------------------------------------------------------ \n" +
"line 20 ------------------------------------------------------------------ \n" +
"line 21 ------------------------------------------------------------------ \n" +
"line 22 ------------------------------------------------------------------ \n" +
"line 23 ------------------------------------------------------------------ \n" +
"line 24 ------------------------------------------------------------------ \n" +
"line 25 ------------------------------------------------------------------ \n" +
"lastline";
			httpClient.start();
			URI uri = URI.create(args[0]);
			httpClient.getAuthenticationStore()
				.addAuthentication(new DigestAuthentication(uri, Authentication.ANY_REALM, "USERNAME", "PASSWORD"));
			ContentResponse response = httpClient.POST(uri)
				.timeout(30, TimeUnit.SECONDS)
				.content(new StringContentProvider(outStr), "text/plain")
				.send();
			System.out.println(response);
		}
		catch (Exception exc) {
			exc.printStackTrace();
		}
		finally {
			httpClient.stop();
		}
    }
}
```

On the client you have to pass in as first parameter the uri pointing to the server.
Maybe you have to increase the size of `outStr` to provocate the failure.

This are samples of Stacktraces of failure
```
java.io.EOFException: HttpConnectionOverHTTP@10334cf(l:/client:52195 <-> r:/remoteserver:8480,closed=false)=>HttpChannelOverHTTP@bd0475(exchange=HttpExchange@1bf9486 req=TERMINATED/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@d7f8f3(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@44e5a2{s=START}],recv=HttpReceiverOverHTTP@c6d773(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]<-SocketChannelEndPoint@1906bc9{/remoteserver:8480<->/client:52195,ISHUT,fill=-,flush=-,to=0/0}{io=0/0,kio=0,kro=1}->HttpConnectionOverHTTP@10334cf(l:/client:52195 <-> r:/remoteserver:8480,closed=false)=>HttpChannelOverHTTP@bd0475(exchange=HttpExchange@1bf9486 req=TERMINATED/null@null res=PENDING/null@null)[send=HttpSenderOverHTTP@d7f8f3(req=QUEUED,snd=COMPLETED,failure=null)[HttpGenerator@44e5a2{s=START}],recv=HttpReceiverOverHTTP@c6d773(rsp=IDLE,failure=null)[HttpParser{s=CLOSED,0 of -1}]]
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.earlyEOF(HttpReceiverOverHTTP.java:320)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1419)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.shutdown(HttpReceiverOverHTTP.java:196)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:143)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:70)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:130)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:116)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:104)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:679)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:597)
	at java.lang.Thread.run(Thread.java:748)
```
```
java.io.IOException: Eine bestehende Verbindung wurde softwaregesteuert
durch den Hostcomputer abgebrochen
	at sun.nio.ch.SocketDispatcher.writev0(Native Method)
	at sun.nio.ch.SocketDispatcher.writev(SocketDispatcher.java:55)
	at sun.nio.ch.IOUtil.write(IOUtil.java:148)
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:504)
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:272)
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:429)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:322)
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:372)
	at org.eclipse.jetty.client.http.HttpSenderOverHTTP$HeadersCallback.process(HttpSenderOverHTTP.java:261)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224)
	at org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(HttpSenderOverHTTP.java:62)
	at org.eclipse.jetty.client.HttpSender.send(HttpSender.java:207)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:83)
	at org.eclipse.jetty.client.HttpConnection.send(HttpConnection.java:202)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:215)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:83)
	at org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(HttpDestinationOverHTTP.java:38)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:332)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:290)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:280)
	at org.eclipse.jetty.client.HttpDestination.release(HttpDestination.java:385)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.release(HttpConnectionOverHTTP.java:137)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.release(HttpChannelOverHTTP.java:89)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.exchangeTerminated(HttpChannelOverHTTP.java:179)
	at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:461)
	at org.eclipse.jetty.client.HttpReceiver.responseSuccess(HttpReceiver.java:405)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.messageComplete(HttpReceiverOverHTTP.java:297)
	at org.eclipse.jetty.http.HttpParser.handleContentMessage(HttpParser.java:607)
	at org.eclipse.jetty.http.HttpParser.parseContent(HttpParser.java:1537)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1385)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:170)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:131)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:70)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:130)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:116)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:104)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
```
org.eclipse.jetty.io.EofException
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:292)
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:429)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:322)
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:372)
	at org.eclipse.jetty.client.http.HttpSenderOverHTTP$HeadersCallback.process(HttpSenderOverHTTP.java:261)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224)
	at org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(HttpSenderOverHTTP.java:62)
	at org.eclipse.jetty.client.HttpSender.send(HttpSender.java:207)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:83)
	at org.eclipse.jetty.client.HttpConnection.send(HttpConnection.java:202)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:215)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:83)
	at org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(HttpDestinationOverHTTP.java:38)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:332)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:290)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:280)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:255)
	at org.eclipse.jetty.client.HttpClient.send(HttpClient.java:557)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:723)
	at org.eclipse.jetty.client.HttpRequest.send(HttpRequest.java:707)
	at org.eclipse.jetty.client.AuthenticationProtocolHandler$AuthenticationListener.onComplete(AuthenticationProtocolHandler.java:176)
	at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:193)
	at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:185)
	at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:459)
	at org.eclipse.jetty.client.HttpReceiver.responseSuccess(HttpReceiver.java:405)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.messageComplete(HttpReceiverOverHTTP.java:297)
	at org.eclipse.jetty.http.HttpParser.handleContentMessage(HttpParser.java:607)
	at org.eclipse.jetty.http.HttpParser.parseContent(HttpParser.java:1537)
	at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1385)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:170)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:131)
	at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:70)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:130)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:116)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:104)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Eine bestehende Verbindung wurde softwaregesteuert
durch den Hostcomputer abgebrochen
	at sun.nio.ch.SocketDispatcher.writev0(Native Method)
	at sun.nio.ch.SocketDispatcher.writev(SocketDispatcher.java:55)
	at sun.nio.ch.IOUtil.write(IOUtil.java:148)
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:504)
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:272)
	... 40 more
```

Thank you for your efforts!
Unfortunately, it does not work yet. I observerd two different problems.

One is that if the result passed to `onComplete` indeed has a request failure but no response failure, the `AuthenticationHandler` does try to resend the request but actually never reaches the server. Maybe it tries to reuse the now-closed connection?

Another problem is that sometimes the result has both, a request failure and a response failure (both pointing to the same exception object, see stacktrace below). I am not sure, but probably this means that this failure happened before the response status was received.
Looking a bit deeper into `HttpClient`, it seems that it only starts receiving the response after the first chunk of content is sent. Provided this is true, if the connection is closed while sending the first chunk, we never receive a response, even if physically it would already be available.
Otherwise, I do not understand how this can be the case.

Here is the stacktrace of the exceptions registered in the result passed to the `AuthenticationHandler`, both for request failures and response failures:
```
org.eclipse.jetty.io.EofException
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:292)
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:429)
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:322)
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:372)
	at org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendContent(HttpSenderOverHTTP.java:106)
	at org.eclipse.jetty.client.HttpSender$ContentCallback.process(HttpSender.java:836)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224)
	at org.eclipse.jetty.client.HttpSender$CommitCallback.process(HttpSender.java:760)
	at org.eclipse.jetty.client.HttpSender$CommitCallback.succeeded(HttpSender.java:697)
	at org.eclipse.jetty.client.http.HttpSenderOverHTTP$HeadersCallback.onCompleteSuccess(HttpSenderOverHTTP.java:318)
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:325)
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:224)
	at org.eclipse.jetty.client.http.HttpSenderOverHTTP.sendHeaders(HttpSenderOverHTTP.java:62)
	at org.eclipse.jetty.client.HttpSender.send(HttpSender.java:207)
	at org.eclipse.jetty.client.http.HttpChannelOverHTTP.send(HttpChannelOverHTTP.java:88)
	at org.eclipse.jetty.client.HttpConnection.send(HttpConnection.java:199)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP$Delegate.send(HttpConnectionOverHTTP.java:254)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.send(HttpConnectionOverHTTP.java:122)
	at org.eclipse.jetty.client.http.HttpDestinationOverHTTP.send(HttpDestinationOverHTTP.java:38)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:332)
	at org.eclipse.jetty.client.HttpDestination.process(HttpDestination.java:290)
	at org.eclipse.jetty.client.HttpDestination.send(HttpDestination.java:280)
	at org.eclipse.jetty.client.HttpDestination.succeeded(HttpDestination.java:221)
	at org.eclipse.jetty.client.AbstractConnectionPool.proceed(AbstractConnectionPool.java:138)
	at org.eclipse.jetty.client.AbstractConnectionPool$1.succeeded(AbstractConnectionPool.java:115)
	at org.eclipse.jetty.client.AbstractConnectionPool$1.succeeded(AbstractConnectionPool.java:108)
	at org.eclipse.jetty.util.Promise$Wrapper.succeeded(Promise.java:130)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onOpen(HttpConnectionOverHTTP.java:130)
	at org.eclipse.jetty.io.SelectorManager.connectionOpened(SelectorManager.java:322)
	at org.eclipse.jetty.io.ManagedSelector.createEndPoint(ManagedSelector.java:213)
	at org.eclipse.jetty.io.ManagedSelector.access$1600(ManagedSelector.java:60)
	at org.eclipse.jetty.io.ManagedSelector$CreateEndPoint.run(ManagedSelector.java:650)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Eine bestehende Verbindung wurde softwaregesteuert
durch den Hostcomputer abgebrochen
	at sun.nio.ch.SocketDispatcher.write0(Native Method)
	at sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:51)
	at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:93)
	at sun.nio.ch.IOUtil.write(IOUtil.java:51)
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:471)
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:270)
```

> I don't think it's related to the Netty issue, since your logs show that we got the exception while reading from the socket, not writing to it.

I am not so sure about that.

I should have mentioned that the above stacktraces are *not* exception stacktraces. Rather, I debugged single-stepping and took two particular snapshots (from the same thread) to show what happens after the following exception is thrown (which I should have included, too):

```
org.eclipse.jetty.io.EofException
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:292)
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:429)
	at org.eclipse.jetty.io.WriteFlusher.completeWrite(WriteFlusher.java:384)
	at org.eclipse.jetty.io.ChannelEndPoint$3.run(ChannelEndPoint.java:139)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:247)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:140)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:382)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:708)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:626)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: Eine bestehende Verbindung wurde softwaregesteuert
durch den Hostcomputer abgebrochen
	at sun.nio.ch.SocketDispatcher.write0(Native Method)
	at sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:51)
	at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:93)
	at sun.nio.ch.IOUtil.write(IOUtil.java:51)
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:471)
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:270)
	... 10 more
```

This is clearly a write failure.

Note that `WriteFlusher.completeWrite(WriteFlusher.java:384)` appears on the stack.
Here, the exception is catched, and inside the `catch` block, `PendingState.fail` is called, which in turn indirectly calls `HttpReceiverOverHTTP.abort`.
Until this point, both stacktraces are identical.

The first stacktrace shows that from here, `HttpParser.close` will be called through `HttpReceiverOverHTTP.dispose`.

The second stacktrace shows that afterwards, `HttpSender.terminateRequest` will be called, which eventually indirectly calls `HttpConnectionOverHTTP.close`. Previously I stopped here, but if we go deeper, we can see that this really closes the underlying channel (again not an exception, but a snapshot):

```
Thread [HttpClient@24377049-12] (Suspended)	
	SocketChannelImpl(AbstractInterruptibleChannel).close() line: 111 [local variables unavailable]	
	SocketChannelEndPoint(ChannelEndPoint).doClose() line: 205	
	SocketChannelEndPoint(AbstractEndPoint).doOnClose(Throwable) line: 212	
	SocketChannelEndPoint(AbstractEndPoint).close(Throwable) line: 184	
	SocketChannelEndPoint(AbstractEndPoint).close() line: 169	
	HttpConnectionOverHTTP.close(Throwable) line: 196	
	HttpConnectionOverHTTP.close() line: 182	
	HttpChannelOverHTTP.exchangeTerminated(HttpExchange, Result) line: 179	
	HttpReceiverOverHTTP(HttpReceiver).terminateResponse(HttpExchange, Result) line: 454	
	HttpReceiverOverHTTP(HttpReceiver).abort(HttpExchange, Throwable) line: 540	
	HttpChannelOverHTTP(HttpChannel).abortResponse(HttpExchange, Throwable) line: 129	
	HttpSenderOverHTTP(HttpSender).terminateRequest(HttpExchange, Throwable, Result) line: 376	
	HttpSenderOverHTTP(HttpSender).abort(HttpExchange, Throwable) line: 561	
	HttpSenderOverHTTP(HttpSender).anyToFailure(Throwable) line: 345	
	HttpSender$ContentCallback.onCompleteFailure(Throwable) line: 897	
	HttpSender$ContentCallback(IteratingCallback).failed(Throwable) line: 401	
	WriteFlusher$PendingState.fail(Throwable) line: 262	
	AbstractEndPoint$2(WriteFlusher).completeWrite() line: 410	
	ChannelEndPoint$3.run() line: 139	
	EatWhatYouKill.doProduce() line: 247	
	EatWhatYouKill.produce() line: 140	
	EatWhatYouKill.run() line: 131	
	ReservedThreadExecutor$ReservedThread.run() line: 382	
	QueuedThreadPool.runJob(Runnable) line: 708	
	QueuedThreadPool$2.run() line: 626	
	Thread.run() line: 745	
```

Thus, as a consequence of the exception thrown while writing, both the parser and the connection are closed.

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java
@@ -140,10 +140,10 @@ public abstract class AuthenticationProtocolHandler implements ProtocolHandler
             }
 
             ContentProvider requestContent = request.getContent();
-            if (requestContent != null && !requestContent.isRepeatable())
+            if (requestContent != null && !requestContent.isReproducible())
             {
                 if (LOG.isDebugEnabled())
-                    LOG.debug("Request content not repeatable for {}", request);
+                    LOG.debug("Request content not reproducible for {}", request);
                 forwardSuccessComplete(request, response);
                 return;
             }

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/AuthenticationProtocolHandler.java
@@ -27,6 +27,7 @@ import java.util.regex.Pattern;
 
 import org.eclipse.jetty.client.api.Authentication;
 import org.eclipse.jetty.client.api.Connection;
+import org.eclipse.jetty.client.api.ContentProvider;
 import org.eclipse.jetty.client.api.ContentResponse;
 import org.eclipse.jetty.client.api.Request;
 import org.eclipse.jetty.client.api.Response;
@@ -86,7 +87,7 @@ public abstract class AuthenticationProtocolHandler implements ProtocolHandler
         {
             HttpRequest request = (HttpRequest)result.getRequest();
             ContentResponse response = new HttpContentResponse(result.getResponse(), getContent(), getMediaType(), getEncoding());
-            if (result.isFailed())
+            if (result.getResponseFailure() != null)
             {
                 if (LOG.isDebugEnabled())
                     LOG.debug("Authentication challenge failed {}", result.getFailure());
@@ -98,7 +99,7 @@ public abstract class AuthenticationProtocolHandler implements ProtocolHandler
             HttpConversation conversation = request.getConversation();
             if (conversation.getAttribute(authenticationAttribute) != null)
             {
-                // We have already tried to authenticate, but we failed again
+                // We have already tried to authenticate, but we failed again.
                 if (LOG.isDebugEnabled())
                     LOG.debug("Bad credentials for {}", request);
                 forwardSuccessComplete(request, response);
@@ -111,7 +112,7 @@ public abstract class AuthenticationProtocolHandler implements ProtocolHandler
             {
                 if (LOG.isDebugEnabled())
                     LOG.debug("Authentication challenge without {} header", header);
-                forwardFailureComplete(request, null, response, new HttpResponseException("HTTP protocol violation: Authentication challenge without " + header + " header", response));
+                forwardFailureComplete(request, result.getRequestFailure(), response, new HttpResponseException("HTTP protocol violation: Authentication challenge without " + header + " header", response));
                 return;
             }
 
@@ -138,9 +139,18 @@ public abstract class AuthenticationProtocolHandler implements ProtocolHandler
                 return;
             }
 
+            ContentProvider requestContent = request.getContent();
+            if (requestContent != null && !requestContent.isRepeatable())
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Request content not repeatable for {}", request);
+                forwardSuccessComplete(request, response);
+                return;
+            }
+
             try
             {
-                final Authentication.Result authnResult = authentication.authenticate(request, response, headerInfo, conversation);
+                Authentication.Result authnResult = authentication.authenticate(request, response, headerInfo, conversation);
                 if (LOG.isDebugEnabled())
                     LOG.debug("Authentication result {}", authnResult);
                 if (authnResult == null)

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/api/ContentProvider.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/api/ContentProvider.java
@@ -22,6 +22,7 @@ import java.io.Closeable;
 import java.nio.ByteBuffer;
 import java.util.Iterator;
 
+import org.eclipse.jetty.client.HttpClient;
 import org.eclipse.jetty.client.util.ByteBufferContentProvider;
 import org.eclipse.jetty.client.util.PathContentProvider;
 
@@ -49,9 +50,17 @@ public interface ContentProvider extends Iterable<ByteBuffer>
     long getLength();
 
     /**
-     * @return whether the content can be provided more than once
+     * <p>Whether this ContentProvider can produce exactly the same content more
+     * than once.</p>
+     * <p>Implementations should return {@code true} only if the content can be
+     * produced more than once, which means that invocations to {@link #iterator()}
+     * must return a new, independent, iterator instance over the content.</p>
+     * <p>The {@link HttpClient} implementation may use this method in particular
+     * cases where it detects that it is safe to retry a request that failed.</p>
+     *
+     * @return whether the content can be produced more than once
      */
-    default boolean isRepeatable()
+    default boolean isReproducible()
     {
         return false;
     }

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/api/ContentProvider.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/api/ContentProvider.java
@@ -49,6 +49,14 @@ public interface ContentProvider extends Iterable<ByteBuffer>
     long getLength();
 
     /**
+     * @return whether the content can be provided more than once
+     */
+    default boolean isRepeatable()
+    {
+        return false;
+    }
+
+    /**
      * An extension of {@link ContentProvider} that provides a content type string
      * to be used as a {@code Content-Type} HTTP header in requests.
      */

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/util/ByteBufferContentProvider.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/util/ByteBufferContentProvider.java
@@ -58,6 +58,12 @@ public class ByteBufferContentProvider extends AbstractTypedContentProvider
     }
 
     @Override
+    public boolean isRepeatable()
+    {
+        return true;
+    }
+
+    @Override
     public Iterator<ByteBuffer> iterator()
     {
         return new Iterator<ByteBuffer>()

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/util/ByteBufferContentProvider.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/util/ByteBufferContentProvider.java
@@ -58,7 +58,7 @@ public class ByteBufferContentProvider extends AbstractTypedContentProvider
     }
 
     @Override
-    public boolean isRepeatable()
+    public boolean isReproducible()
     {
         return true;
     }
@@ -90,12 +90,6 @@ public class ByteBufferContentProvider extends AbstractTypedContentProvider
                 {
                     throw new NoSuchElementException();
                 }
-            }
-
-            @Override
-            public void remove()
-            {
-                throw new UnsupportedOperationException();
             }
         };
     }

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/util/BytesContentProvider.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/util/BytesContentProvider.java
@@ -54,7 +54,7 @@ public class BytesContentProvider extends AbstractTypedContentProvider
     }
 
     @Override
-    public boolean isRepeatable()
+    public boolean isReproducible()
     {
         return true;
     }
@@ -83,12 +83,6 @@ public class BytesContentProvider extends AbstractTypedContentProvider
                 {
                     throw new NoSuchElementException();
                 }
-            }
-
-            @Override
-            public void remove()
-            {
-                throw new UnsupportedOperationException();
             }
         };
     }

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/util/BytesContentProvider.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/util/BytesContentProvider.java
@@ -54,6 +54,12 @@ public class BytesContentProvider extends AbstractTypedContentProvider
     }
 
     @Override
+    public boolean isRepeatable()
+    {
+        return true;
+    }
+
+    @Override
     public Iterator<ByteBuffer> iterator()
     {
         return new Iterator<ByteBuffer>()

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/util/PathContentProvider.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/util/PathContentProvider.java
@@ -85,6 +85,12 @@ public class PathContentProvider extends AbstractTypedContentProvider
         return fileSize;
     }
 
+    @Override
+    public boolean isRepeatable()
+    {
+        return true;
+    }
+
     public ByteBufferPool getByteBufferPool()
     {
         return bufferPool;

--- a/jetty-client/src/main/java/org/eclipse/jetty/client/util/PathContentProvider.java
+++ b/jetty-client/src/main/java/org/eclipse/jetty/client/util/PathContentProvider.java
@@ -86,7 +86,7 @@ public class PathContentProvider extends AbstractTypedContentProvider
     }
 
     @Override
-    public boolean isRepeatable()
+    public boolean isReproducible()
     {
         return true;
     }

--- a/jetty-client/src/test/java/org/eclipse/jetty/client/AbstractHttpClientServerTest.java
+++ b/jetty-client/src/test/java/org/eclipse/jetty/client/AbstractHttpClientServerTest.java
@@ -27,6 +27,7 @@ import org.eclipse.jetty.server.Handler;
 import org.eclipse.jetty.server.Server;
 import org.eclipse.jetty.server.ServerConnector;
 import org.eclipse.jetty.toolchain.test.TestTracker;
+import org.eclipse.jetty.util.SocketAddressResolver;
 import org.eclipse.jetty.util.ssl.SslContextFactory;
 import org.eclipse.jetty.util.thread.QueuedThreadPool;
 import org.junit.After;
@@ -96,6 +97,7 @@ public abstract class AbstractHttpClientServerTest
         clientThreads.setName("client");
         client = new HttpClient(transport, sslContextFactory);
         client.setExecutor(clientThreads);
+        client.setSocketAddressResolver(new SocketAddressResolver.Sync());
         client.start();
     }
 

--- a/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java
+++ b/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java
@@ -22,11 +22,14 @@ import java.io.File;
 import java.io.IOException;
 import java.net.URI;
 import java.nio.ByteBuffer;
-import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.IntFunction;
 
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServletRequest;
@@ -34,6 +37,7 @@ import javax.servlet.http.HttpServletResponse;
 
 import org.eclipse.jetty.client.api.Authentication;
 import org.eclipse.jetty.client.api.AuthenticationStore;
+import org.eclipse.jetty.client.api.ContentProvider;
 import org.eclipse.jetty.client.api.ContentResponse;
 import org.eclipse.jetty.client.api.Request;
 import org.eclipse.jetty.client.api.Response;
@@ -41,6 +45,7 @@ import org.eclipse.jetty.client.api.Result;
 import org.eclipse.jetty.client.util.BasicAuthentication;
 import org.eclipse.jetty.client.util.DeferredContentProvider;
 import org.eclipse.jetty.client.util.DigestAuthentication;
+import org.eclipse.jetty.http.HttpStatus;
 import org.eclipse.jetty.security.Authenticator;
 import org.eclipse.jetty.security.ConstraintMapping;
 import org.eclipse.jetty.security.ConstraintSecurityHandler;
@@ -220,7 +225,7 @@ public class HttpClientAuthenticationTest extends AbstractHttpClientServerTest
             {
                 baseRequest.setHandled(true);
                 if (requests.incrementAndGet() == 1)
-                    response.sendRedirect(URIUtil.newURI(scheme,request.getServerName(),request.getServerPort(),request.getRequestURI(),null));
+                    response.sendRedirect(URIUtil.newURI(scheme, request.getServerName(), request.getServerPort(), request.getRequestURI(), null));
             }
         });
 
@@ -259,7 +264,7 @@ public class HttpClientAuthenticationTest extends AbstractHttpClientServerTest
             {
                 baseRequest.setHandled(true);
                 if (request.getRequestURI().endsWith("/redirect"))
-                    response.sendRedirect(URIUtil.newURI(scheme,request.getServerName(),request.getServerPort(),"/secure",null));
+                    response.sendRedirect(URIUtil.newURI(scheme, request.getServerName(), request.getServerPort(), "/secure", null));
             }
         });
 
@@ -425,6 +430,40 @@ public class HttpClientAuthenticationTest extends AbstractHttpClientServerTest
     }
 
     @Test
+    public void test_NonRepeatableContent() throws Exception
+    {
+        startBasic(new EmptyServerHandler());
+
+        AuthenticationStore authenticationStore = client.getAuthenticationStore();
+        URI uri = URI.create(scheme + "://localhost:" + connector.getLocalPort());
+        BasicAuthentication authentication = new BasicAuthentication(uri, realm, "basic", "basic");
+        authenticationStore.addAuthentication(authentication);
+
+        CountDownLatch resultLatch = new CountDownLatch(1);
+        byte[] data = new byte[]{'h', 'e', 'l', 'l', 'o'};
+        DeferredContentProvider content = new DeferredContentProvider(ByteBuffer.wrap(data))
+        {
+            @Override
+            public boolean isRepeatable()
+            {
+                return false;
+            }
+        };
+        Request request = client.newRequest(uri)
+                .path("/secure")
+                .content(content);
+        request.send(result ->
+        {
+            if (result.isSucceeded() && result.getResponse().getStatus() == HttpStatus.UNAUTHORIZED_401)
+                resultLatch.countDown();
+        });
+
+        content.close();
+
+        Assert.assertTrue(resultLatch.await(5, TimeUnit.SECONDS));
+    }
+
+    @Test
     public void test_RequestFailsAfterResponse() throws Exception
     {
         startBasic(new EmptyServerHandler());
@@ -434,32 +473,111 @@ public class HttpClientAuthenticationTest extends AbstractHttpClientServerTest
         BasicAuthentication authentication = new BasicAuthentication(uri, realm, "basic", "basic");
         authenticationStore.addAuthentication(authentication);
 
-        CountDownLatch successLatch = new CountDownLatch(1);
+        AtomicBoolean fail = new AtomicBoolean(true);
+        GeneratingContentProvider content = new GeneratingContentProvider(index ->
+        {
+            switch (index)
+            {
+                case 0:
+                    return ByteBuffer.wrap(new byte[]{'h', 'e', 'l', 'l', 'o'});
+                case 1:
+                    return ByteBuffer.wrap(new byte[]{'w', 'o', 'r', 'l', 'd'});
+                case 2:
+                    if (fail.compareAndSet(true, false))
+                    {
+                        // Wait for the 401 response to arrive
+                        // to the authentication protocol handler.
+                        sleep(1000);
+                        // Trigger request failure.
+                        throw new RuntimeException();
+                    }
+                    else
+                    {
+                        return null;
+                    }
+                default:
+                    throw new IllegalStateException();
+            }
+        });
         CountDownLatch resultLatch = new CountDownLatch(1);
-        DeferredContentProvider content = new DeferredContentProvider();
-        Request request = client.newRequest("localhost", connector.getLocalPort())
+        client.newRequest("localhost", connector.getLocalPort())
                 .scheme(scheme)
                 .path("/secure")
                 .content(content)
-                .onResponseSuccess(response -> successLatch.countDown());
-        request.send(result ->
-        {
-            if (result.isFailed() && result.getResponseFailure() == null)
-                resultLatch.countDown();
-        });
+                .send(result ->
+                {
+                    if (result.isSucceeded() && result.getResponse().getStatus() == HttpStatus.OK_200)
+                        resultLatch.countDown();
+                });
 
-        // Send some content to make sure the request is dispatched on the server.
-        content.offer(ByteBuffer.wrap("hello".getBytes(StandardCharsets.UTF_8)));
-
-        // Wait for the response to arrive to
-        // the authentication protocol handler.
-        Thread.sleep(1000);
-
-        // Trigger request failure.
-        request.abort(new Exception());
-
-        // Verify that the response was successful, it's the request that failed.
-        Assert.assertTrue(successLatch.await(5, TimeUnit.SECONDS));
         Assert.assertTrue(resultLatch.await(5, TimeUnit.SECONDS));
+    }
+
+    private void sleep(long time)
+    {
+        try
+        {
+            Thread.sleep(time);
+        }
+        catch (InterruptedException x)
+        {
+            throw new RuntimeException(x);
+        }
+    }
+
+    private static class GeneratingContentProvider implements ContentProvider
+    {
+        private static final ByteBuffer DONE = ByteBuffer.allocate(0);
+
+        private final IntFunction<ByteBuffer> generator;
+
+        private GeneratingContentProvider(IntFunction<ByteBuffer> generator)
+        {
+            this.generator = generator;
+        }
+
+        @Override
+        public long getLength()
+        {
+            return -1;
+        }
+
+        @Override
+        public boolean isRepeatable()
+        {
+            return true;
+        }
+
+        @Override
+        public Iterator<ByteBuffer> iterator()
+        {
+            return new Iterator<ByteBuffer>()
+            {
+                private int index;
+                public ByteBuffer current;
+
+                @Override
+                public boolean hasNext()
+                {
+                    if (current == null)
+                    {
+                        current = generator.apply(index++);
+                        if (current == null)
+                            current = DONE;
+                    }
+                    return current != DONE;
+                }
+
+                @Override
+                public ByteBuffer next()
+                {
+                    ByteBuffer result = current;
+                    current = null;
+                    if (result == null)
+                        throw new NoSuchElementException();
+                    return result;
+                }
+            };
+        }
     }
 }

--- a/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java
+++ b/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java
@@ -444,7 +444,7 @@ public class HttpClientAuthenticationTest extends AbstractHttpClientServerTest
         DeferredContentProvider content = new DeferredContentProvider(ByteBuffer.wrap(data))
         {
             @Override
-            public boolean isRepeatable()
+            public boolean isReproducible()
             {
                 return false;
             }
@@ -543,7 +543,7 @@ public class HttpClientAuthenticationTest extends AbstractHttpClientServerTest
         }
 
         @Override
-        public boolean isRepeatable()
+        public boolean isReproducible()
         {
             return true;
         }

--- a/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java
+++ b/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientAuthenticationTest.java
@@ -430,7 +430,7 @@ public class HttpClientAuthenticationTest extends AbstractHttpClientServerTest
     }
 
     @Test
-    public void test_NonRepeatableContent() throws Exception
+    public void test_NonReproducibleContent() throws Exception
     {
         startBasic(new EmptyServerHandler());
 

